packages/aztec.js/src/proof/bilateralSwap/helpers.js

```diff
@@ -0,0 +1,170 @@
+const BN = require('bn.js');
+const crypto = require('crypto');
+
+
+const bn128 = require('../../bn128');
+const secp256k1 = require('../../secp256k1');
+const notesConstruct = require('../../note');
+
+
+const { groupReduction } = bn128;
+
+
+const helpers = {};
+
+helpers.makeTestNotes = (makerNoteValues, takerNoteValues) => {
+    const noteValues = makerNoteValues.concat(takerNoteValues);
+    const numNotes = noteValues.length;
+
+    let i;
+    const publicKeys = [];
+    for (i = 0; i < numNotes; i += 1) {
+        const { publicKey } = secp256k1.accountFromPrivateKey(crypto.randomBytes(32));
+        publicKeys.push(publicKey);
+    }
+
+    const testNotes = publicKeys.map((publicKey, j) => {
+        return notesConstruct.create(publicKey, noteValues[j]);
+    });
+
+    return testNotes;
+};
+
+helpers.checkNumberNotes = (notes, numberExpected) => {
+    if (notes.length !== numberExpected) {
+        throw new Error('Incorrect number of notes');
+    }
+};
+
+helpers.makeNoteArray = (notes) => {
+    const makerNotes = Object.values(notes.makerNotes);
+    const takerNotes = Object.values(notes.takerNotes);
+    const noteArray = [makerNotes[0], makerNotes[1], takerNotes[0], takerNotes[1]];
+    return noteArray;
+};
+
+helpers.makeIncorrectArray = (notes) => {
+    const makerNotes = Object.values(notes.makerNotes);
+    const takerNotes = Object.values(notes.takerNotes);
+    const noteArray = [makerNotes[0], makerNotes[1], makerNotes[2], takerNotes[0], takerNotes[1], takerNotes[2]];
+    return noteArray;
+};
+
+helpers.validateOnCurve = (x, y) => {
+    const rhs = x.redSqr().redMul(x).redAdd(bn128.curve.b);
+    const lhs = y.redSqr();
+    if (!rhs.fromRed().eq(lhs.fromRed())) {
+        throw new Error('point not on the curve');
+    }
+};
+
+helpers.getBlindingFactorsAndChallenge = (noteArray, finalHash) => {
+    const bkArray = [];
+    const blindingFactors = noteArray.map((note, i) => {
+        let bk = bn128.randomGroupScalar();
+        const ba = bn128.randomGroupScalar();
+        let B;
+
+        /*
+        Explanation of the below if/else
+        - The purpose is to set bk1 = bk3 and bk2 = bk4
+        - i is used as an indexing variable, to keep track of whether we are at a maker note or taker note
+        - All bks are stored in a bkArray. When we arrive at the taker notes, we set bk equal to the bk of the corresponding 
+          maker note. This is achieved by 'jumping back' 2 index positions (i - 2) in the bkArray, and setting the current
+          bk equal to the element at the resulting position.
+        */
+
+        // Maker notes
+        if (i <= 1) {
+            B = note.gamma.mul(bk).add(bn128.h.mul(ba));
```

Comment:
Computation of `B` should be outside the `if, else` block

Reply:
Done